clang/runtime/dpct-rt/include/sparse_utils.hpp.inc

```diff
@@ -78,6 +79,241 @@ private:
+    oneapi::mkl::sparse::gemm(queue, oneapi::mkl::layout::row_major, trans,
+                              oneapi::mkl::transpose::nontrans, alpha_value,
+                              *sparse_matrix_handle, data_b, dense_cols, ldb,
+                              beta_value, data_c, ldc);
+    break;
+  }
+  default:
+    throw std::runtime_error(
+        "the csrmm does not support matrix_info::matrix_type::sy, "
+        "matrix_info::matrix_type::tr amd matrix_info::matrix_type::he");
+  }
+
+  sycl::event e =
+      oneapi::mkl::sparse::release_matrix_handle(queue, sparse_matrix_handle);
+  queue.submit([&](sycl::handler &cgh) {
+    cgh.depends_on(e);
+    cgh.host_task([=] { delete sparse_matrix_handle; });
+  });
+}
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|optimize_info|dpct::sparse
+// DPCT_DEPENDENCY_EMPTY
+// DPCT_CODE
+/// Saving the optimization information for solving a system of linear
+/// equations.
+class optimize_info {
+public:
+  /// Constructor
+  optimize_info() { oneapi::mkl::sparse::init_matrix_handle(&_matrix_handle); }
+  /// Destructor
+  ~optimize_info() {
+    oneapi::mkl::sparse::release_matrix_handle(get_default_queue(),
+                                               &_matrix_handle, _deps)
+        .wait();
+  }
+  /// Add dependency for the destructor.
+  /// \param [in] e The event which the destructor depends on.
+  void add_dependency(sycl::event e) { _deps.push_back(e); }
+  /// Get the internal saved matrix handle.
+  /// \return Returns the matrix handle.
+  oneapi::mkl::sparse::matrix_handle_t get_matrix_handle() const noexcept {
+    return _matrix_handle;
+  }
+
+private:
+  oneapi::mkl::sparse::matrix_handle_t _matrix_handle = nullptr;
+  std::vector<sycl::event> _deps;
+};
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|optimize_csrsv|dpct::sparse
+// DPCT_DEPENDENCY_BEGIN
+// SparseUtils|matrix_info
+// SparseUtils|optimize_info
+// Util|DataType
+// LibCommonUtils|get_memory
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+/// Performs internal optimizations for solving a system of linear equations for
+/// a CSR format sparse matrix.
+/// \param [in] queue The queue where the routine should be executed.
+/// \param [in] trans The operation applied to the sparse matrix.
+/// \param [in] row_col Number of rows of the sparse matrix.
+/// \param [in] info Matrix info of the sparse matrix.
+/// \param [in] val An array containing the non-zero elements of the sparse matrix.
+/// \param [in] row_ptr An array of length \p num_rows + 1.
+/// \param [in] col_ind An array containing the column indices in index-based
+/// numbering.
+/// \param [out] optimize_info The result of the optimizations.
+template <typename T>
+void optimize_csrsv(sycl::queue &queue, oneapi::mkl::transpose trans,
+                    int row_col, const std::shared_ptr<matrix_info> info,
+                    const T *val, const int *row_ptr, const int *col_ind,
+                    std::shared_ptr<optimize_info> optimize_info) {
```

Comment:
I see the optimize_csrsv api which analyzes the sparse matrix graph to extract parallelism,  but don't see the corresponding csrsv api which cna perform the trsv operation.  Is that to be added later ?

Reply:
Yes. This PR focus on to refine old migrations.
`csrsv` helper function will be added in the future.